pkg/estimator/server/framework/runtime/framework.go

```diff
@@ -0,0 +1,130 @@
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/informers"
+
+	"github.com/karmada-io/karmada/pkg/estimator/pb"
+	"github.com/karmada-io/karmada/pkg/estimator/server/framework"
+	"github.com/karmada-io/karmada/pkg/estimator/server/metrics"
+	utilmetrics "github.com/karmada-io/karmada/pkg/util/metrics"
+)
+
+const (
+	estimator = "Estimator"
+)
+
+// frameworkImpl implements the Framework interface and is responsible for initializing and running scheduler
+// plugins.
+type frameworkImpl struct {
+	estimateReplicasPlugins []framework.EstimateReplicasPlugin
+	informerFactory         informers.SharedInformerFactory
+}
+
+var _ framework.Framework = &frameworkImpl{}
+
+type frameworkOptions struct {
+	informerFactory informers.SharedInformerFactory
+}
+
+// Option for the frameworkImpl.
+type Option func(*frameworkOptions)
+
+func defaultFrameworkOptions() frameworkOptions {
+	return frameworkOptions{}
+}
+
+// WithInformerFactory sets informer factory for the scheduling frameworkImpl.
+func WithInformerFactory(informerFactory informers.SharedInformerFactory) Option {
+	return func(o *frameworkOptions) {
+		o.informerFactory = informerFactory
+	}
+}
+
+// NewFramework creates a scheduling framework by registry.
+func NewFramework(r Registry, opts ...Option) (framework.Framework, error) {
+	options := defaultFrameworkOptions()
+	for _, opt := range opts {
+		opt(&options)
+	}
+	f := &frameworkImpl{
+		informerFactory: options.informerFactory,
+	}
+	estimateReplicasPluginsList := reflect.ValueOf(&f.estimateReplicasPlugins).Elem()
+	estimateReplicasType := estimateReplicasPluginsList.Type().Elem()
+
+	for name, factory := range r {
+		p, err := factory(f)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize plugin %q: %w", name, err)
+		}
+		addPluginToList(p, estimateReplicasType, &estimateReplicasPluginsList)
+	}
+	return f, nil
+}
+
+func addPluginToList(plugin framework.Plugin, pluginType reflect.Type, pluginList *reflect.Value) {
+	if reflect.TypeOf(plugin).Implements(pluginType) {
+		newPlugins := reflect.Append(*pluginList, reflect.ValueOf(plugin))
+		pluginList.Set(newPlugins)
+	}
+}
+
+// SharedInformerFactory returns a shared informer factory.
+func (frw *frameworkImpl) SharedInformerFactory() informers.SharedInformerFactory {
+	return frw.informerFactory
+}
+
+// RunEstimateReplicasPlugins runs the set of configured EstimateReplicasPlugins
+// for estimating replicas based on the given replicaRequirements.
+// It returns an integer and an error.
+// The integer represents the minimum calculated value of estimated replicas from each EstimateReplicasPlugin.
+func (frw *frameworkImpl) RunEstimateReplicasPlugins(ctx context.Context, replicaRequirements *pb.ReplicaRequirements) (int32, error) {
+	startTime := time.Now()
+	defer func() {
+		metrics.FrameworkExtensionPointDuration.WithLabelValues(estimator).Observe(utilmetrics.DurationInSeconds(startTime))
+	}()
+	var result int32 = math.MaxInt32
```

Comment:
If no plugins exist, we should return zero?

Reply:
If no plugins, the replica is calculated using node resources. If set to zero when not plugins, then the the final estimation will be zero even if there are enough resources to run the job (we compare the results and return the smaller one ?).

If no plugins, return math.MaxInt32 means no operation.

Reply:
> If we set to zero, then the the final estimation will be zero even if there are enough resources to run the job (we compare the results and return the smaller one ?

That's what I have mentioned above. We'd better transform estimation by node idle resources to a new plugin. Along with resource plugin and quota plugin, we decide the final estimation result. Does it make sense?

Reply:
I prefer to think estimation by node idle resources as the "core" and plugin is for add-on features and more light-weight.

If we really want to convert that to NodeResourceEstimator as default feature, I prefer to add `*scheduler.Cache` in the Handle and keep the `RunEstimateReplicasPlugins` as
```
RunEstimateReplicasPlugins(ctx context.Context, replicaRequirements *pb.ReplicaRequirements) (int32, error)
```

Reply:
I think more about it. It might be useful to change estimation by node idle resource as a plugin because I hope to introduce eviction by priority class. 

My native think about that is to write a plugin to estimate the replica based on node idle resources and resources taken by pods' resources with lower priority. I will turn off the default estimation by node idle resource. 
The pod eviction is handled by the assigned clusters' k8s scheduler.

Reply:
@Garrybest, @RainbowMango and @chaosi-zju if we agree to change the current estimation by idle resource to the default plugin. Could we keep RunEstimateReplicasPlugins as it is now

`RunEstimateReplicasPlugins(ctx context.Context, replicaRequirements *pb.ReplicaRequirements) (int32, error)`

and add `*scheduler.Cache` in the Handle as I explained in ?
https://github.com/karmada-io/karmada/pull/4534#discussion_r1456891393

Reply:
Snapshot from a cache is a little bit expensive. Different plugins could not share a snapshot if make cache as a part of handler cause every plugin should do the snapshot work if needed, right?

Reply:
It is so true, we need to put snapshot as an input arg
How about 
```
Estimate(ctx context.Context, snapshot *schedcache.Snapshot, replicaRequirements *pb.ReplicaRequirements) (int32, error)
```
make it similar
 [estimateReplicas](https://github.com/karmada-io/karmada/blob/master/pkg/estimator/server/estimate.go#L58-L61 )?

@Garrybest and @RainbowMango and @chaosi-zju do you think if I should convert the current [estimateReplicas](https://github.com/karmada-io/karmada/blob/master/pkg/estimator/server/estimate.go#L58-L61 ) to a default plugin or make that change in another PR ?

Reply:
Need @Garrybest 's point here. I'm not sure if it's the original idea or just makes filter and score plugins.
https://github.com/karmada-io/karmada/blob/697170b90463176a519b19f9e4eab22acd07e8a0/pkg/estimator/server/estimate.go#L76

Both are OK to me for now. 
Keeping the current estimation as a `core` makes sense because we can go back to consider this after the resource quota plugin gets onboard.
Moving the current estimation to the plugin also makes sense because it `seems` can support more use cases (just as mentioned https://github.com/karmada-io/karmada/pull/4534#discussion_r1456914573 above). But, we need more effort for it ,such as make plugin configurable.

But, please don't do it in this PR. (Keep each small and we can move faster)

Reply:
Exactly, I think we just add snapshot as an input and optimize resource plugin in next PR. This PR generally looks good.

Reply:
Got it.

Reply:
> Snapshot from a cache is a little bit expensive. 

Good point! I'm sold.